File: Delphi-source-code/MyFinance/MyFinancePersonnel/src/uFinancePersonnelBL.pas
unit uFinancePersonnelBL;

interface

uses
  System.SysUtils, System.Classes;

const
  SQL_LABOR_COST_REPORT =
    'SELECT e.EmployeeId, e.FirstName + '' '' + e.LastName AS Name, e.Position, ' +
    'e.Salary AS BaseSalary, b.BenefitsCost AS Benefits, ' +
    'e.Salary + b.BenefitsCost AS TotalCost ' +
    'FROM Employees e LEFT JOIN EmployeeBenefits b ON e.EmployeeId = b.EmployeeId ' +
    'WHERE e.Active = 1 ORDER BY TotalCost DESC';

  SQL_OVERTIME_REPORT =
    'SELECT t.EmployeeId, e.FirstName + '' '' + e.LastName AS Name, ' +
    'SUM(t.RegularHours) AS RegularHours, SUM(t.OvertimeHours) AS OvertimeHours, ' +
    'SUM(t.OvertimeHours * e.Salary / 2080 * 1.5) AS OvertimeCost ' +
    'FROM Timesheets t INNER JOIN Employees e ON t.EmployeeId = e.EmployeeId ' +
    'WHERE t.Period = :Period GROUP BY t.EmployeeId, e.FirstName, e.LastName';

  SQL_HEADCOUNT_BY_POSITION =
    'SELECT e.Position, COUNT(*) AS HeadCount, AVG(e.Salary) AS AvgSalary ' +
    'FROM Employees e WHERE e.Active = 1 ' +
    'GROUP BY e.Position ORDER BY HeadCount DESC';

type
  TLaborCostRecord = record
    EmployeeId: Integer;
    Name: string;
    Position: string;
    BaseSalary: Currency;
    Benefits: Currency;
    TotalCost: Currency;
  end;

  TOvertimeRecord = record
    EmployeeId: Integer;
    Name: string;
    RegularHours: Double;
    OvertimeHours: Double;
    OvertimeCost: Currency;
  end;

  THeadcountRecord = record
    Position: string;
    Count: Integer;
    AvgSalary: Currency;
  end;

  TFinancePersonnelBL = class
  public
    function GetLaborCostReport: TArray<TLaborCostRecord>;
    function GetOvertimeReport: TArray<TOvertimeRecord>;
    function GetHeadcountByPosition: TArray<THeadcountRecord>;
  end;

implementation

{ TFinancePersonnelBL }

function TFinancePersonnelBL.GetLaborCostReport: TArray<TLaborCostRecord>;
begin
  SetLength(Result, 9);

  Result[0].EmployeeId := 9;
  Result[0].Name := 'Robert Johnson';
  Result[0].Position := 'General Manager';
  Result[0].BaseSalary := 62000;
  Result[0].Benefits := 12400;
  Result[0].TotalCost := 74400;

  Result[1].EmployeeId := 1;
  Result[1].Name := 'Maria Garcia';
  Result[1].Position := 'Head Chef';
  Result[1].BaseSalary := 55000;
  Result[1].Benefits := 11000;
  Result[1].TotalCost := 66000;

  Result[2].EmployeeId := 2;
  Result[2].Name := 'Carlos Reyes';
  Result[2].Position := 'Sous Chef';
  Result[2].BaseSalary := 42000;
  Result[2].Benefits := 8400;
  Result[2].TotalCost := 50400;

  Result[3].EmployeeId := 3;
  Result[3].Name := 'Ana Martinez';
  Result[3].Position := 'Floor Manager';
  Result[3].BaseSalary := 38000;
  Result[3].Benefits := 7600;
  Result[3].TotalCost := 45600;

  Result[4].EmployeeId := 4;
  Result[4].Name := 'James Wilson';
  Result[4].Position := 'Bartender';
  Result[4].BaseSalary := 32000;
  Result[4].Benefits := 6400;
  Result[4].TotalCost := 38400;

  Result[5].EmployeeId := 5;
  Result[5].Name := 'Sofia Lopez';
  Result[5].Position := 'Waiter';
  Result[5].BaseSalary := 28000;
  Result[5].Benefits := 5600;
  Result[5].TotalCost := 33600;

  Result[6].EmployeeId := 6;
  Result[6].Name := 'David Chen';
  Result[6].Position := 'Waiter';
  Result[6].BaseSalary := 28000;
  Result[6].Benefits := 5600;
  Result[6].TotalCost := 33600;

  Result[7].EmployeeId := 7;
  Result[7].Name := 'Emily Brown';
  Result[7].Position := 'Host';
  Result[7].BaseSalary := 26000;
  Result[7].Benefits := 5200;
  Result[7].TotalCost := 31200;

  Result[8].EmployeeId := 8;
  Result[8].Name := 'Miguel Torres';
  Result[8].Position := 'Dishwasher';
  Result[8].BaseSalary := 24000;
  Result[8].Benefits := 4800;
  Result[8].TotalCost := 28800;
end;

function TFinancePersonnelBL.GetOvertimeReport: TArray<TOvertimeRecord>;
begin
  SetLength(Result, 8);

  Result[0].EmployeeId := 1;
  Result[0].Name := 'Maria Garcia';
  Result[0].RegularHours := 160;
  Result[0].OvertimeHours := 12;
  Result[0].OvertimeCost := 475.96;

  Result[1].EmployeeId := 2;
  Result[1].Name := 'Carlos Reyes';
  Result[1].RegularHours := 160;
  Result[1].OvertimeHours := 18;
  Result[1].OvertimeCost := 544.23;

  Result[2].EmployeeId := 3;
  Result[2].Name := 'Ana Martinez';
  Result[2].RegularHours := 160;
  Result[2].OvertimeHours := 8;
  Result[2].OvertimeCost := 219.23;

  Result[3].EmployeeId := 4;
  Result[3].Name := 'James Wilson';
  Result[3].RegularHours := 160;
  Result[3].OvertimeHours := 22;
  Result[3].OvertimeCost := 507.69;

  Result[4].EmployeeId := 5;
  Result[4].Name := 'Sofia Lopez';
  Result[4].RegularHours := 152;
  Result[4].OvertimeHours := 6;
  Result[4].OvertimeCost := 121.15;

  Result[5].EmployeeId := 6;
  Result[5].Name := 'David Chen';
  Result[5].RegularHours := 160;
  Result[5].OvertimeHours := 10;
  Result[5].OvertimeCost := 201.92;

  Result[6].EmployeeId := 7;
  Result[6].Name := 'Emily Brown';
  Result[6].RegularHours := 144;
  Result[6].OvertimeHours := 0;
  Result[6].OvertimeCost := 0;

  Result[7].EmployeeId := 8;
  Result[7].Name := 'Miguel Torres';
  Result[7].RegularHours := 160;
  Result[7].OvertimeHours := 15;
  Result[7].OvertimeCost := 259.62;
end;

function TFinancePersonnelBL.GetHeadcountByPosition: TArray<THeadcountRecord>;
begin
  SetLength(Result, 7);

  Result[0].Position := 'Waiter';
  Result[0].Count := 2;
  Result[0].AvgSalary := 28000;

  Result[1].Position := 'Head Chef';
  Result[1].Count := 1;
  Result[1].AvgSalary := 55000;

  Result[2].Position := 'Sous Chef';
  Result[2].Count := 1;
  Result[2].AvgSalary := 42000;

  Result[3].Position := 'Floor Manager';
  Result[3].Count := 1;
  Result[3].AvgSalary := 38000;

  Result[4].Position := 'Bartender';
  Result[4].Count := 1;
  Result[4].AvgSalary := 32000;

  Result[5].Position := 'Host';
  Result[5].Count := 1;
  Result[5].AvgSalary := 26000;

  Result[6].Position := 'Dishwasher';
  Result[6].Count := 1;
  Result[6].AvgSalary := 24000;
end;

end.

File: MyFinancePersonnel.Core/Application/Services/MyFinancePersonnel/uFinancePersonnelBL.cs
using Microsoft.Extensions.Logging;
using MyFinancePersonnel.Core.Application.DTOs.Core;

namespace MyFinancePersonnel.Core.Application.Services.MyFinancePersonnel;

public class uFinancePersonnelBL : IuFinancePersonnelBL
{
    private readonly ILogger<uFinancePersonnelBL> _logger;

    public uFinancePersonnelBL(ILogger<uFinancePersonnelBL> logger)
    {
        _logger = logger;
    }

    public List<TLaborCostRecord> GetLaborCostReport()
    {
        _logger.LogInformation("Generating labor cost report");

        return new List<TLaborCostRecord>
        {
            new TLaborCostRecord
            {
                EmployeeId = 9,
                Name = "Robert Johnson",
                Position = "General Manager",
                BaseSalary = 62000m,
                Benefits = 12400m,
                TotalCost = 74400m
            },
            new TLaborCostRecord
            {
                EmployeeId = 1,
                Name = "Maria Garcia",
                Position = "Head Chef",
                BaseSalary = 55000m,
                Benefits = 11000m,
                TotalCost = 66000m
            },
            new TLaborCostRecord
            {
                EmployeeId = 2,
                Name = "Carlos Reyes",
                Position = "Sous Chef",
                BaseSalary = 42000m,
                Benefits = 8400m,
                TotalCost = 50400m
            },
            new TLaborCostRecord
            {
                EmployeeId = 3,
                Name = "Ana Martinez",
                Position = "Floor Manager",
                BaseSalary = 38000m,
                Benefits = 7600m,
                TotalCost = 45600m
            },
            new TLaborCostRecord
            {
                EmployeeId = 4,
                Name = "James Wilson",
                Position = "Bartender",
                BaseSalary = 32000m,
                Benefits = 6400m,
                TotalCost = 38400m
            },
            new TLaborCostRecord
            {
                EmployeeId = 5,
                Name = "Sofia Lopez",
                Position = "Waiter",
                BaseSalary = 28000m,
                Benefits = 5600m,
                TotalCost = 33600m
            },
            new TLaborCostRecord
            {
                EmployeeId = 6,
                Name = "David Chen",
                Position = "Waiter",
                BaseSalary = 28000m,
                Benefits = 5600m,
                TotalCost = 33600m
            },
            new TLaborCostRecord
            {
                EmployeeId = 7,
                Name = "Emily Brown",
                Position = "Host",
                BaseSalary = 26000m,
                Benefits = 5200m,
                TotalCost = 31200m
            },
            new TLaborCostRecord
            {
                EmployeeId = 8,
                Name = "Miguel Torres",
                Position = "Dishwasher",
                BaseSalary = 24000m,
                Benefits = 4800m,
                TotalCost = 28800m
            }
        };
    }

    public async Task<List<TLaborCostRecord>> GetLaborCostReportAsync(CancellationToken ct = default)
    {
        return await Task.FromResult(GetLaborCostReport());
    }

    public List<TOvertimeRecord> GetOvertimeReport()
    {
        _logger.LogInformation("Generating overtime report");

        return new List<TOvertimeRecord>
        {
            new TOvertimeRecord
            {
                EmployeeId = 1,
                Name = "Maria Garcia",
                RegularHours = 160.0,
                OvertimeHours = 12.0,
                OvertimeCost = 475.96m
            },
            new TOvertimeRecord
            {
                EmployeeId = 2,
                Name = "Carlos Reyes",
                RegularHours = 160.0,
                OvertimeHours = 18.0,
                OvertimeCost = 544.23m
            },
            new TOvertimeRecord
            {
                EmployeeId = 3,
                Name = "Ana Martinez",
                RegularHours = 160.0,
                OvertimeHours = 8.0,
                OvertimeCost = 219.23m
            },
            new TOvertimeRecord
            {
                EmployeeId = 4,
                Name = "James Wilson",
                RegularHours = 160.0,
                OvertimeHours = 22.0,
                OvertimeCost = 507.69m
            },
            new TOvertimeRecord
            {
                EmployeeId = 5,
                Name = "Sofia Lopez",
                RegularHours = 152.0,
                OvertimeHours = 6.0,
                OvertimeCost = 121.15m
            },
            new TOvertimeRecord
            {
                EmployeeId = 6,
                Name = "David Chen",
                RegularHours = 160.0,
                OvertimeHours = 10.0,
                OvertimeCost = 201.92m
            },
            new TOvertimeRecord
            {
                EmployeeId = 7,
                Name = "Emily Brown",
                RegularHours = 144.0,
                OvertimeHours = 0.0,
                OvertimeCost = 0m
            },
            new TOvertimeRecord
            {
                EmployeeId = 8,
                Name = "Miguel Torres",
                RegularHours = 160.0,
                OvertimeHours = 15.0,
                OvertimeCost = 259.62m
            }
        };
    }

    public async Task<List<TOvertimeRecord>> GetOvertimeReportAsync(CancellationToken ct = default)
    {
        return await Task.FromResult(GetOvertimeReport());
    }

    public List<THeadcountRecord> GetHeadcountByPosition()
    {
        _logger.LogInformation("Generating headcount by position report");

        return new List<THeadcountRecord>
        {
            new THeadcountRecord
            {
                Position = "Waiter",
                Count = 2,
                AvgSalary = 28000m
            },
            new THeadcountRecord
            {
                Position = "Head Chef",
                Count = 1,
                AvgSalary = 55000m
            },
            new THeadcountRecord
            {
                Position = "Sous Chef",
                Count = 1,
                AvgSalary = 42000m
            },
            new THeadcountRecord
            {
                Position = "Floor Manager",
                Count = 1,
                AvgSalary = 38000m
            },
            new THeadcountRecord
            {
                Position = "Bartender",
                Count = 1,
                AvgSalary = 32000m
            },
            new THeadcountRecord
            {
                Position = "Host",
                Count = 1,
                AvgSalary = 26000m
            },
            new THeadcountRecord
            {
                Position = "Dishwasher",
                Count = 1,
                AvgSalary = 24000m
            }
        };
    }

    public async Task<List<THeadcountRecord>> GetHeadcountByPositionAsync(CancellationToken ct = default)
    {
        return await Task.FromResult(GetHeadcountByPosition());
    }
}
